proposals/NNNN-swift-noasync.md

```diff
@@ -0,0 +1,329 @@
+    _ mutex: UnsafeMutablePointer<pthread_mutex_t>,
+    do op: () throws -> R) rethrows -> R {
+  switch pthread_mutex_lock(mutex) {
+    case 0:
+      defer { pthread_mutex_unlock(mutex) }
+      return try op()
+    case EINVAL:
+      preconditionFailure("Invalid Mutex")
+    case EDEADLK:
+      fatalError("Locking would cause a deadlock")
+    case let value:
+      fatalError("Unknown pthread_mutex_lock() return value: '\(value)'")
+  }
+}
+```
+
+The above snippet is a safe wrapper for `pthread_mutex_lock` and
+`pthread_mutex_unlock`, since the lock is not held across suspension points. The
+critical section operation must be synchronous for this to hold true though.
+The following snippet uses a synchronous closure to call the unavailable
+function, circumventing the protection provided by the attribute.
+
+```swift
+@available(*, noasync)
+func pthread_mutex_lock(_ lock: UnsafeMutablePointer<pthread_mutex_t>) {}
+
+func asyncFun(_ mutex : UnsafeMutablePointer<pthread_mutex_t>) async {
+  // Error! pthread_mutex_lock is unavailable from async contexts
+  pthread_mutex_lock(mutex)
+
+  // Ok! pthread_mutex_lock is not called from an async context
+  _ = { unavailableFun(mutex) }()
+
+  await someAsyncOp()
+}
+```
+
+### Replacement API
+
+In some cases, it is possible to provide an alternative that is safe. The
+`with_pthread_mutex_lock` is an example of a way to provide a safe way to wrap
+locking and unlocking pthread mutexes.
+
+In other cases, it may be safe to use an API from a specific executor. For
+example, API that uses thread-local storage isn't safe for consumption by
+asynchronous functions in general, but is safe for functions on the MainActor
+since it will only use the main thread.
+
+The unavailable API should still be annotated as such, but an alternative
+function can be implemented as an extension of an actor.
+
+```swift
+@available(*, noasync, renamed: "mainactorReadID()", message: "use mainactorReadID instead")
+func readIDFromThreadLocal() -> Int { }
+
+@MainActor
+func readIDFromMainActor() -> Int { readIDFromThreadLocal() }
+
+func asyncFunc() async {
+  // Bad, we don't know what thread we're on
+  let id = readIDFromThreadLocal()
+
+  // Good, we know it's coming from the main actor on the main thread.
+  // Note that we have to jump to the main actor, so there is a suspension.
+  let id = await readIDFromMainActor()
+}
+```
+
+Custom executors are a goal for the language in the future. Restricting an API
+to a custom executor is achieved in the same way that restricting the API to a
+global actor is done. In many ways, they are the same thing. An actor ensures
+that methods and operations on that actor are done on the executor for that
+actor. By setting the `unownedExecutor` of an actor to the desired executor and
+creating a wrapper function for the restricted API in that actor, the restricted
+API is now only available for use from asynchronous contexts on that actor, and
+in extension, that executor.
```

Comment:
Is the example below meant to show an `IOActor` with a custom executor and how that interacts with `noasync`? If so, maybe this discussion (and the example) should be a Future Directions sort of section. The example should also have a comment saying to assume that it's not an ordinary actor, etc.

Reply:
Yeah, I was going back-and-forth on whether it belonged in a `future work` section. It's not really a future direction of this proposal, but just how it will interact with future proposals since it came up in the discussion on the forum.

Reply:
Done